File: test/ModularPipelines.UnitTests/Dependencies/ModuleCategoryAttributeTests.cs
using System.Reflection;
using ModularPipelines.Attributes;
using ModularPipelines.TestHelpers;

namespace ModularPipelines.UnitTests.Dependencies;

public class ModuleCategoryAttributeTests : TestBase
{
    public async Task Constructor_WithValidCategory_SetsCategoryProperty()
    {
        var attr = new ModuleCategoryAttribute("infrastructure");
        await Assert.That(attr.Category).IsEqualTo("infrastructure");
    }

    public async Task Constructor_WithNullCategory_ThrowsArgumentException()
    {
        await Assert.That(() => new ModuleCategoryAttribute(null!))
            .ThrowsException()
            .And.IsTypeOf<ArgumentException>();
    }

    public async Task Constructor_WithEmptyCategory_ThrowsArgumentException()
    {
        await Assert.That(() => new ModuleCategoryAttribute(""))
            .ThrowsException()
            .And.IsTypeOf<ArgumentException>();
    }

    public async Task Attribute_DoesNotAllowMultiple()
    {
        var usage = typeof(ModuleCategoryAttribute).GetCustomAttribute<AttributeUsageAttribute>();
        await Assert.That(usage!.AllowMultiple).IsFalse();
    }

    public async Task Attribute_IsInheritable()
    {
        var usage = typeof(ModuleCategoryAttribute).GetCustomAttribute<AttributeUsageAttribute>();
        await Assert.That(usage!.Inherited).IsTrue();
    }
}

File: test/ModularPipelines.UnitTests.FSharp/Dependencies/ModuleCategoryAttributeTests.fs
namespace ModularPipelines.UnitTests.FSharp.Dependencies

open ModularPipelines.UnitTests.Dependencies
open ModularPipelines.UnitTests.FSharp
open TUnit.Core

type ModuleCategoryAttributeTests() =
    inherit ModularPipelines.UnitTests.Dependencies.ModuleCategoryAttributeTests()

    [<Test>]
    member this.Constructor_WithValidCategory_SetsCategoryProperty() =
        CSharpTestWrapper.invokeTest this typeof<ModularPipelines.UnitTests.Dependencies.ModuleCategoryAttributeTests> "Constructor_WithValidCategory_SetsCategoryProperty" None

    [<Test>]
    member this.Constructor_WithNullCategory_ThrowsArgumentException() =
        CSharpTestWrapper.invokeTest this typeof<ModularPipelines.UnitTests.Dependencies.ModuleCategoryAttributeTests> "Constructor_WithNullCategory_ThrowsArgumentException" None

    [<Test>]
    member this.Constructor_WithEmptyCategory_ThrowsArgumentException() =
        CSharpTestWrapper.invokeTest this typeof<ModularPipelines.UnitTests.Dependencies.ModuleCategoryAttributeTests> "Constructor_WithEmptyCategory_ThrowsArgumentException" None

    [<Test>]
    member this.Attribute_DoesNotAllowMultiple() =
        CSharpTestWrapper.invokeTest this typeof<ModularPipelines.UnitTests.Dependencies.ModuleCategoryAttributeTests> "Attribute_DoesNotAllowMultiple" None

    [<Test>]
    member this.Attribute_IsInheritable() =
        CSharpTestWrapper.invokeTest this typeof<ModularPipelines.UnitTests.Dependencies.ModuleCategoryAttributeTests> "Attribute_IsInheritable" None
